Vectors Layer

bryantChhun opened the pull request:
# Description
<!-- What does this pull request (PR) do? Why is it necessary? -->
This PR includes a new layer to display vectors.  It also includes demo scripts and some tests.
<!-- Tell us about your new feature, improvement, or fix! -->
Building off of the markers layer as a guide, the vectors layer is able to render lines and accepts two data formats:
1) image-like per-pixel format of shape (N, M, 2)
- each position of the NxM array defines the position of the vector.
- each element of the NxM array contains (x-projection, y-projection) of the vector
- the length of the projection is not checked and can extend into nearby pixels
2) arbitrary coordinates as an array of shape (N, 4)
- element of the array contains (x-position, y-position, x-projection, y-projection) of the vector

Three new layer menu items are available:
1) Averaging:
- adjusts the underlying data to resample the vectors by a (i, j) pre-defined kernel
- default implementation works only for image-like data
- ability to bind new functions to this GUI selection so that custom compute can adjust the underlying data

2) Length:
- adjusts the underlying data to extend the length of all vectors
- ability to bind new functions to this GUI selection so that custom compute can adjust the underlying data

3) Width:
- adjusts the thickness of the lines

This PR reflects only 2D vector layer rendering (3D can be a later improvement)

## Type of change
<!-- Please delete options that are not relevant. -->
- [ ] Bug-fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# References
<!-- What resources, documentation, and guides were used in the creation of this PR? -->
Lots of descriptive doc strings.  Two examples serve as decent guides for how to use the layer.
<!-- If this is a bug-fix or otherwise resolves an issue, reference it here with "closes #(issue)" -->

# How has this been tested?
<!-- Please describe the tests that you ran to verify your changes. -->
- [x] example: the test suite for my feature covers cases x, y, and z
I test that the layer handles correct and incorrect data formats (by throwing the custom exception "InvalidDataFormatError")
I test that vector data can be assigned to the layer after construction
I test that custom behavior (averaging and length callbacks) perform as expected
- [x] example: all tests pass with my change
All tests pass.
I could write yet more tests that probe data more data shapes, such as 3d data, and proper slicing of them.  But this would be for a later PR that includes 3d rendering.

## Final checklist:
- [x] My PR is the minimum possible work for the desired functionality
yes? :-)
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works

bryantChhun committed:
Merge pull request #1 from Napari/master

sync with napari/napari-gui master

bryantChhun committed:
Merge pull request #2 from Napari/master

Sync my fork with master

bryantChhun committed:
Merge pull request #3 from Napari/master

note on readme about stability added (#76)

bryantChhun committed:
Merge pull request #5 from Napari/master

include qt layer refactor

bryantChhun committed:
Fixed layer

bryantChhun committed:
Fixed layer

bryantChhun committed:
merged

bryantChhun committed:
merged

bryantChhun committed:
merged

bryantChhun committed:
final version of vectors layer

bryantChhun committed:
final version of vectors layer

bryantChhun committed:
demo to show vectors layer, script mode only

bryantChhun committed:
demo to show vectors layer, script mode only

bryantChhun committed:
removed signals
settled on length widget type

bryantChhun committed:
removed signals
settled on length widget type

bryantChhun committed:
added more documentation
removed signals and settled callback method

bryantChhun committed:
added more documentation
removed signals and settled callback method

bryantChhun committed:
docstrings for clarity.
adjusted lines and markers

bryantChhun committed:
docstrings for clarity.
adjusted lines and markers

bryantChhun committed:
removed commented code

bryantChhun committed:
removed commented code

bryantChhun committed:
test init

bryantChhun committed:
test init

bryantChhun committed:
towards new data model

bryantChhun committed:
renamed some

bryantChhun committed:
refactor to add data type:
layer can now receive data of format (x, y, u, v)

bryantChhun committed:
Vector layer now supports both data types
examples for each vector layer data type

bryantChhun committed:
improved demos

bryantChhun committed:
adjusted for data types, length and averaging callbacks

bryantChhun committed:
default average settled, requires cv2 dependency

bryantChhun committed:
few changes to matrix and data

bryantChhun committed:
bigger matrix

bryantChhun committed:
Merge branch 'master' of https://github.com/Napari/napari-gui into Vectors

# Conflicts:
#	examples/demo.py
#	examples/layers.ipynb
#	gui/layers/qt/__init__.py

bryantChhun committed:
vector layer for new refactor

bryantChhun committed:
vector layer for new refactor

bryantChhun committed:
more interesting pattern to demonstrate power of coordinate lines

bryantChhun committed:
vector layer tests

bryantChhun committed:
more interesting data

bryantChhun committed:
cleaned up commented code

bryantChhun committed:
more comments
added _current_shape to track image shape
removed unused, old code

bryantChhun committed:
removed commented code

bryantChhun committed:
Another example with constantly updating vectors

bryantChhun committed:
removed comment

bryantChhun committed:
removed cv2 dependency

bryantChhun committed:
pep8

bryantChhun committed:
adjusted to reflect new example format

sofroniewn commented on .idea/inspectionProfiles/Project_Default.xml:
```diff
@@ -0,0 +1,15 @@
+<component name="InspectionProjectProfileManager">
```
Can you add the `.idea` folder to the `.gitignore` or remove it from the repo. It is not clear what it is doing here.

bryantChhun replied:
isn't .gitignore a versioned file?  I looked up some ways to unversion .idea without adding it to .gitignore, so as not to pollute .gitignore.  But if you're ok with it, I can add .idea there.

btw, .idea files are generated by Pycharm.  Surely, I can't be the only one using Pycharm..?

sofroniewn replied:
I have no objection to adding `.idea` files / folders to the `.gitignore`

bryantChhun replied:
removed and added .idea to gitignore.

sofroniewn commented on .idea/inspectionProfiles/Project_Default.xml:
```diff
@@ -0,0 +1,15 @@
+<component name="InspectionProjectProfileManager">
+  <profile version="1.0">
```
Similar to above

sofroniewn commented on examples/demo_mult_image.py:
```diff
@@ -0,0 +1,51 @@
+#!/usr/bin/env python
```
Can you remove this autogenerated header info from this and similar files

bryantChhun replied:
removed this demo as it was not related to vectors

sofroniewn commented on examples/layers.ipynb:
```diff
@@ -9,7 +9,7 @@
   },
```
can you remove the `layers.ipynb` and replace with additional script examples if necessary

sofroniewn commented on examples/markers.ipynb:
```diff
@@ -2,9 +2,17 @@
  "cells": [
```
can you remove the `markers.ipynb` and replace with additional script examples if necessary

sofroniewn commented on examples/tests/__init__.py:
```diff
@@ -0,0 +1,9 @@
+#!/usr/bin/env python
```
tests should probably not live inside a folder in examples. Not sure where they should live, (either temporarily or in the long run). Maybe @kne42 or @AhmetCanSolak want's to weigh in

bryantChhun replied:
will move once i sync up with loic's changes

sofroniewn commented on examples/demo_mult_image.py:
```diff
@@ -0,0 +1,51 @@
+#!/usr/bin/env python
+# title           : this_python_file.py
```
In this example (and others like it) can you drop the word `demo` from the title

sofroniewn commented on gui/layers/_vectors_layer/model.py:
```diff
@@ -0,0 +1,624 @@
+
+from .._base_layer import Layer
+from .._register import add_to_viewer
+from gui._vispy.scene.visuals import Line as LinesNode
```
This import should be a relative path

`from ..._vispy.scene.visuals import Line as LinesNode`

sofroniewn commented on gui/layers/_vectors_layer/model.py:
```diff
@@ -0,0 +1,624 @@
+        averaging.setter adjusts the underlying data
+        subscribe an observer by registering it with "averaging_bind_to"
+
+    width : int
```
As width increases, I encounter a point where the lines no longer increase in width (depending on my zoom). Do other's notice this too? I think this is a vispy (and possibly open-gl) problem under the hood, but just wanted to note here

bryantChhun replied:
Interesting.  I am able to increase the width to about maximum 65 pixels in both demos.  Above that, the rendered image doesn't change..

sofroniewn replied:
Hmm, mine still stops only around 3 pixels. It's a bug in the vispy stuff I think. I've seen some warnings about it in their docs. I have some alternative code for rendering lines, but I'm not sure we need to go down that path yet. Depends on how annoying people find it. Worth making a note in the docs maybe

sofroniewn commented on gui/layers/_vectors_layer/model.py:
```diff
@@ -0,0 +1,624 @@
+    width : int
+        width of the line in pixels
+
+    length : int or float
```
Not obvious to me what this does in all examples. I'm looking at the `demo_vectors_coord-like.py` example and changing length doesn't seem to change anything. Can you explain?

bryantChhun replied:
I enable length and averaging only for image-like data.  Averaging only makes sense for image-like data where (n,m) kernels must all contain some value.  Additionally, length and averaging adjust the underlying data, which i think makes most sense with the image-like data.

sofroniewn replied:
In that case can you set the `averaging` and `length` ComboBoxes to be `disabled` if `self._data_type` is `'projection'`. This will cause them to be grayed out and non-interactive

sofroniewn commented on gui/layers/_vectors_layer/view/controls.py:
```diff
@@ -0,0 +1,124 @@
+                self.layer.mode = 'pan/zoom'
+
+
+class QtSelectButton(QRadioButton):
```
Can probably delete this as not in use right now. Do you have plans / desires for interactivity (line drawing etc.) or would that come in the shapes layer?

bryantChhun replied:
don't have current plans but thought it'd be nice.  This can easily be added at a later date

sofroniewn commented on gui/layers/_vectors_layer/view/controls.py:
```diff
@@ -0,0 +1,124 @@
+                self.layer.mode = 'select'
+
+
+class QtAdditionButton(QRadioButton):
```
Similar to comment above, delete as not in use?

sofroniewn commented on gui/layers/_vectors_layer/view/properties.py:
```diff
@@ -0,0 +1,81 @@
+        self.grid_layout.addWidget(QLabel('width:'), 4, 0)
+        self.grid_layout.addWidget(width_field, 4, 1)
+
+        # averaging combobox
```
When I am using `demo_vectors_coord-like` this does not seem to have any effect - why is that? Is that expected?

sofroniewn replied:
When I use `demo_vectors_image-like` it behaves as expected. If there is a fundamental different between the image like and the coord like we should think about how to best expose this interactivity.

bryantChhun replied:
Same as above

sofroniewn commented on gui/layers/_vectors_layer/view/properties.py:
```diff
@@ -0,0 +1,81 @@
+        self.grid_layout.addWidget(QLabel('averaging:'), 5, 0)
+        self.grid_layout.addWidget(averaging_combobox, 5, 1)
+
+        # line length
```
When I am using `demo_vectors_coord-like` this does not seem to have any effect - why is that? Is that expected?

sofroniewn replied:
Similar to above, when I use `demo_vectors_image-like` it behaves as expected. If there is a fundamental different between the image like and the coord like we should think about how to best expose this interactivity.

sofroniewn commented on gui/layers/_vectors_layer/view/properties.py:
```diff
@@ -0,0 +1,81 @@
+        self.grid_layout.addWidget(QLabel('color:'), 3, 0)
+        self.grid_layout.addWidget(face_comboBox, 3, 1)
+
+        # line width in pixels
```
Is there a reason you prefer the `QComboBox` to the `QSlider` in this case? Could have `QSlider` and a `QTextBox`?

bryantChhun replied:
If you're referring to the "color" widget, i just copied the implementation from markers layer.
If you're referring to width, then yes I could try a slider.  Thing is that the documentation here says width is in pixels (http://api.vispy.org/en/latest/visuals.html#vispy.visuals.LineVisual).  Generally I like the precision of the numbers over slider.
Also, I did try slider for the length widget, but it would call too many updates.  And because length adjusts the data, the interaction would slow down dramatically.

sofroniewn replied:
I was referring to the width. I am fine with the `QSpinBox`, maybe I will change it in the markers / shapes layers too for consistency. I take your point that it is nice to have that precision

sofroniewn commented on examples/demo_vectors_coord-like.py:
```diff
@@ -0,0 +1,41 @@
+"""
```
Can you rename this example `add_vectors_coords` and as a minor edit make the text at the top look a little more like the text at the top of that `add_images` and `add_markers` examples (i.e. replace `This example ... ` with `Display ...`)

sofroniewn commented on examples/demo_vectors_image-like.py:
```diff
@@ -0,0 +1,38 @@
+"""
```
Similar to above, can you rename this example `add_vectors_image` and as a minor edit make the text at the top look a little more like the text at the top of that add_images and add_markers examples (i.e. replace This example ... with Display ...)

sofroniewn commented on examples/demo_mult_image.py:
```diff
@@ -0,0 +1,51 @@
+#!/usr/bin/env python
+# title           : this_python_file.py
+# description     :This will create a header for a python script.
```
This example seems redundant with the `add_images` example which adds two images, and also doesn't contain anything specific to the vectors. Can you remove from this PR

bryantChhun replied:
removed

sofroniewn commented on examples/demo_mult_image_sockets.py:
```diff
@@ -0,0 +1,97 @@
+#!/usr/bin/env python
```
This example seems to have some performance issues on my computer (pan/zoom become slow) and it is not quite clear to me what it does. Doesn't seem related to vectors, and also seems to be a bit to advanced functionality (i.e. custom Qt Signals/Slots for our current examples. Maybe you keep a copy of this locally and then in the future if there is more of an advanced use-case section we could highlight this.

sofroniewn replied:
Running it again I see the timed updates, but still think this feature doesn't belong in the examples at this point

bryantChhun replied:
also removed

sofroniewn commented on examples/demo_mult_image_sockets2.py:
```diff
@@ -0,0 +1,144 @@
+#!/usr/bin/env python
```
Similar to my comment on the other example with sockets this functionality seems too advanced for our examples right now. It was cool to watch the animation, and in the future we can think about more animation / multiprocess examples, but I think you should maintain this separately for now

kne42 commented on examples/demo_mult_image_sockets2.py:
```diff
@@ -0,0 +1,144 @@
+import numpy as np
+import time
+
+"""
```
docstring should go at the top of the file

bryantChhun replied:
removing this for now

bryantChhun committed:
added .idea to gitignore

bryantChhun committed:
renamed

bryantChhun committed:
fixed headers
fixed relative imports
removed unused controls

bryantChhun committed:
fixed headers
fixed relative imports
removed unused controls

bryantChhun committed:
removed .idea files

sofroniewn commented on gui/layers/_vectors_layer/model.py:
```diff
@@ -0,0 +1,615 @@
+    _avg_dims
+    _averaging
+    _length
+    _avg_observers
```
Why do you have the `_avg_observers` and `_len_observers`? Could this instead be accomplished with the event based implementation that we put on properties in #155. In this case setting the `averaging` or `length` properties would trigger the `averaging` or `length` events that could be hooked up to whatever other things you wanted. If so I'd prefer you moving over to the event style for consistency

bryantChhun replied:
Yes this is totally possible.  I remember Loic also mentioning a listener module at the last meeting.  Events could do the trick too.

The reason these methods exist is because they adjust the underlying data.  The literal way i've done vector averaging (in _default_avg and _default_length) doesn't always make sense depending on the application/data.  Registering an observer allows one to pay attention to user-GUI inputs and have external code adjust the data in another, meaningful way.

bryantChhun replied:
removed this observer pattern in favor of vispy's event system.  There is no longer any tracking of observers in a list within the vector_layer.

I moved the averaging and length adjustment math to the qt properties, but am open to switching this.  My reasoning is that qt properties contains slots, while the layer appears to contain emitters, and the averaging/length functions behave mostly as a slot.

sofroniewn commented on gui/layers/_vectors_layer/model.py:
```diff
@@ -0,0 +1,615 @@
+        self._width = width
+        self._color = color
+        self._colors = get_color_names()
+        self._connector_types = ['segments']
```
Is it really necessary to support the `strips` connector type? Can you not get away with hard-coding the `segments` connector type. Given how the initial data gets set in either the `matrix` or `projection` type it doesn't seem like you need `strips` or am I mistaken

bryantChhun replied:
No it's not necessary to have strips connector type nor ndarray.  I've been thinking about hard coding that, but keep second guessing myself.

It would seriously confuse things too, if we would allow the user to adjust the way vertices are connected.  Would effectively turn this layer into much more than a vectors layer.

I'll make the change

sofroniewn replied:
Sounds good

sofroniewn commented on gui/layers/_vectors_layer/model.py:
```diff
@@ -0,0 +1,615 @@
+        self._kernel = self._kernel_dict['1x1']
+
+        # Store underlying data model
+        self._data_types = ('matrix', 'projection')
```
Why do you call the data types `matrix` and `projection` here, but then in the top of the docs and the examples call them `image-like` and `coordinate-like`. I'd prefer consistency and lean towards `image` and `coords`.

sofroniewn commented on gui/layers/_vectors_layer/model.py:
```diff
@@ -0,0 +1,615 @@
+        self._refresh()
+
+    @property
+    def mode(self):
```
It's unclear to me if we need any of the mode stuff. Note that the `image` layer doesn't have it, and I feel like if you aren't going to have multiple buttons on the left hand side then there is no point. I'd probably delete it (and just delete the `controls.py` inside view too. I can't remember now what happens if the `_qt_controls` property is `None`, but I think it just get's ignored and that space is left blank (if that doesn't happen we should fix it so it does).

Unless there is something else you'd rather put on the left hand side like a slider? Or you can imagine more interactive buttons etc

bryantChhun replied:
would it make sense to have averaging and length adjustments on the left side?  Or is the left side mostly for buttons and/or sliders?

I'm totally fine with deleting controls completely.

sofroniewn replied:
I think we should just delete the controls completely for now. Can always add them back in if we see a need for special interactivity, but I think how you have the averaging and length adjustment with the other properties makes more sense right now

sofroniewn commented on gui/layers/_vectors_layer/model.py:
```diff
@@ -0,0 +1,615 @@
+        super().__init__(visual)
+
+        # map averaging type to tuple
+        self._kernel_dict = {'1x1': (1, 1),
```
Thinking about this more - why do you not just allow for arbitrary kernel sizes (or at least restrict to odd ones with a 2n+1 mapping). Having the list seems unnecessarily restrictive. I suggest having an `_kernel` property that is either a single integer (in which case it gets copied across all dimensions) or is a tuple (with the same length as dimensions). In the GUI I would then expose a `QSpinBox` like you do for length, and as that is adjusted it adjusts the `_kernel` to be the integer represented their (possibly after the 2n+1 transform), so that the `QSpinBox` would show `0` for no averaging.

jni replied:
Exactly. This is one of many spots where I think an nd generalisation would be trivial, and actually result in cleaner code.

bryantChhun replied:
Ok, I've changed this so that it takes a single QSpinbox and allows arbitrary 2n+1 values.  

It would be nice to have two spinboxes side by side, one for each (x,y) kernel.  But assigning two columns to a single row of grid_layout will alter all other rows.  I tried adding a layout to the specific row of grid_layout, but then it causes "hide" problems ...

sofroniewn replied:
Nice! My concern about having the two spinboxes is that it won't generalize to ND images. I'd say for now we don't worry about it and try and get this merged, and then we can revisit this when we think about the nD generalization

sofroniewn commented on gui/layers/_vectors_layer/model.py:
```diff
@@ -0,0 +1,615 @@
+        self.viewer._child_layer_changed = True
+        self._refresh()
+
+    def _check_vector_type(self, vectors):
```
I feel like this function does more than just a check - it actually also returns the `cood_list` that gets set to the `_vectors` property. Can you rename accordingly

sofroniewn commented on gui/layers/_vectors_layer/model.py:
```diff
@@ -0,0 +1,615 @@
+        # Get a list of the vectors for the vectors in this slice
+
+        vectors = self.vectors
+        if len(vectors) > 0:
```
RIght now this logic is making certain assumptions about how nD Vectors data would be formatted, but you don't really support nD data at all. I'd suggest just deleting / commenting out this whole function for now and when it comes time to make this layer type be nD we can implement something new. It will likely want to incorporate the refactoring of the dims that @royerloic is working on right now, so best to wait on this

sofroniewn commented on gui/layers/_vectors_layer/model.py:
```diff
@@ -0,0 +1,615 @@
+        indices : sequence of int or slice
+            Indices to slice with.
+        """
+        in_slice_vectors, matches = self._slice_vectors(indices)
```
If you see my comment bellow about deleting `_slice_vectors`, I would just remove this line and have
`in_slice_vectors = self.vectors`

sofroniewn replied:
Doing this will mean that they layer will not crash if you try and overlay the 2D vectors layer on a 3D or 4D image, instead the 2D vectors will always appear regardless of what slice of the image you are looking at, which I think is fine for now, and we can work on an nD vectors layer after the new dims refactoring

sofroniewn commented on gui/layers/_vectors_layer/view/controls.py:
```diff
@@ -0,0 +1,68 @@
+
+from PyQt5.QtWidgets import QButtonGroup, QVBoxLayout, QRadioButton, QFrame
```
See my comment around `set_mode` on if this whole file can just be deleted

sofroniewn commented on gui/layers/_vectors_layer/model.py:
```diff
@@ -0,0 +1,615 @@
+        self.name = 'vectors'
+        self.events.add(mode=Event)
+        self._qt_properties = QtVectorsLayer(self)
+        self._qt_controls = QtVectorsControls(self)
```
See my comment around `set_mode` on whether you can delete this line too. `_qt_controls` is set to `None` in the base class, and I just checked that everything will launch fine without it

sofroniewn reviewed (changes requested):
I have gone through the PR again and think we are close to having it ready to merge once my last set of comments are addressed.

The most important one is how we approach the nD stuff. I think we should wait for now, have this be a purely 2D layer type, and then after @royerloic merges the new dims stuff we can add support for nD options

bryantChhun committed:
Merge pull request #7 from napari/master

sync my fork

bryantChhun committed:
master is synched with Vectors
master did pull from bryant's fork, this commit reflects the vector changes

bryantChhun committed:
Merge branch 'master' of https://github.com/bryantChhun/napari-gui

bryantChhun committed:
Merge branch 'master' of https://github.com/bryantChhun/napari-gui into Vectors

bryantChhun committed:
Merge branch 'master' of https://github.com/bryantChhun/napari-gui into Vectors

bryantChhun committed:
Remove .idea files from commit

bryantChhun committed:
Merge branch 'master' of https://github.com/bryantChhun/napari-gui into Vectors

bryantChhun committed:
removed vector controls and unnecessary header

bryantChhun committed:
added name parameter in constructor
removed endpoint from linspace that caused pixel centering problems (for image-like vectors only)

bryantChhun committed:
dramatically sped up default_averaging by reducing loops and using \
np.ndenumerate on strided indices

bryantChhun committed:
enabled fractions of length adjustments.
DoubleSpinBox now increments by 0.1 per click

bryantChhun committed:
included scipy dependency
use scipy.signal.2dconvolve to speed up averaging

bryantChhun committed:
fixed scaling at higher averaging kernels

bryantChhun committed:
add_vectors_image: larger array and adjustment to sample data to include negative values
model: Now uses the vispy Event system upon "averaging" and "length" GUI adjustments.  Removed observer pattern in favor of vispy's emitter system
view.properties: now contains averaging and length adjustment math
tests: fixed to reflect above changes

sofroniewn commented on napari/layers/_vectors_layer/model.py:
```diff
@@ -0,0 +1,461 @@
+    _width
+    _color
+    _colors
+    _connector_types
```
I'd remove this and have the `connector` hard coded to be `segments`

sofroniewn commented on napari/layers/_vectors_layer/view/properties.py:
```diff
@@ -0,0 +1,125 @@
+    def change_length(self, value):
+        self.layer.length = value
+
+    def _default_avg(self, event_kernel):
```
My thoughts are that all this averaging should not live in the properties file, but in the model file. It is more consistent with the rest of the layers to do it this way. Curious what @royerloic thinks

bryantChhun replied:
I've changed it now.  I placed it there because I saw it as a slot and wanted to stay consistent with where slots and signals existed.  But it's simple to make another properties file method that calls the layer._default_avg -- so I've done this

bryantChhun replied:
there's another concern I have here, which might be fixable:

as it exists, the flow looks like this:
1) user changes averaging in UI properties
2) layer.averaging property is directly called and set
3) layer.averaging property setter emits a signal
4) UI receives the signal and calls another UI method
5) this other UI method calls layer._default_avg

So now we have a situation where the calls bounce back and forth between layer and properties.  But it is necessitated by the fact that I want observers for this UI event, and I want to silo where signals and slots exist in a manner consistent with how properties/layers are written.

sofroniewn replied:
yup, I think we will likely revisit this when we refactor our layers (like how we are refactoring our dims), but for now it is consistent with the other layer types. Can you want to post this concern as an issue that we can then discuss more.

sofroniewn commented on napari/layers/_vectors_layer/model.py:
```diff
@@ -0,0 +1,461 @@
+    ----------
+    Private attributes
+    -------
+    _kernel_dict
```
this is no longer present and needs to be removed

sofroniewn commented on napari/layers/_vectors_layer/model.py:
```diff
@@ -0,0 +1,461 @@
+
+        self._refresh()
+
+    def averaging_bind_to(self, callback):
```
Do we really need to provide this binding here? We don't do it for our other properties in our other layers, and it is not something that this layer uses by default. I think there will be a bigger conversation about how do we trigger plugin code based on events that happen in the layer (like when a shape is clicked on run a computation) but we havn't started that yet, and I'd rather merge this now without these bindings and solve these issues then

bryantChhun replied:
Maybe up for discussion as I do see this as very useful to me.  It might be useful to the larger community?  It enables someone who defines "averaging" of vectors differently to implement their own calculation.  We could make it private if that helps?

sofroniewn replied:
Why don't you just overwrite the averaging method directly, i.e.

```python
def myavgmethod():
    print('new method')

vect._default_avg = myavgmethod
```

to me this is much cleaner than having the `averaging_bind_to` function and something that only fairly advanced users will want to do. I just gave it a try and it works fine. I'd rename `_default_avg` to something more like `_averaging_method`

bryantChhun replied:
ah yes... this makes sense and is embarrassingly obvious, especially if we would make the above method private anyway.  Will do.

sofroniewn commented on napari/layers/_vectors_layer/model.py:
```diff
@@ -0,0 +1,461 @@
+
+        self._refresh()
+
+    def length_bind_to(self, callback):
```
see my comment about adding the averaging binding - i think the same applies here too

sofroniewn commented on napari/layers/_vectors_layer/view/properties.py:
```diff
@@ -0,0 +1,125 @@
+            self.layer.vectors = output_mat[x_offset:range_x-x_offset:x, y_offset:range_y-y_offset:y]
+            self.layer.length = self.layer._length
+
+    def _default_length(self, event_len):
```
I'd also put this in the main model file if i understand it correctly - it only modifies code in the model

sofroniewn reviewed (changes requested):
@bryantChhun looking really good - i just made a couple comments. I think it's close to being ready

bryantChhun committed:
removed averaging math from properties (placed in layer)
hard code connector type
cleaned up documentation
removed _kernel as it is redundant with _averaging

bryantChhun committed:
removed method to register external observers for UI events (averaging and length)

bryantChhun committed:
removed tests

bryantChhun committed:
reformatted for pytest using app context

bryantChhun committed:
cleaned up the example

bryantChhun committed:
removed tests.  Problems with CI and were written as integration tests not unittests

sofroniewn committed:
Merge branch 'master' into Vectors

kne42 commented on examples/notebook.ipynb:
```diff
@@ -16,7 +16,7 @@
   {
    "cell_type": "code",
-   "execution_count": null,
+   "execution_count": 1,
```
should reset this

kne42 commented on napari/layers/_vectors_layer/view/properties.py:
```diff
@@ -0,0 +1,85 @@
+        for c in colors:
+            face_comboBox.addItem(c)
+        index = face_comboBox.findText(self.layer.color, Qt.MatchFixedString)
+        if index >= 0:
```
what's this for? when is the index <=0 ?

sofroniewn replied:
This happens if the color is not found in the list of supported colors, and is something that should probably be explained in the markers and shapes layer too

kne42 commented on napari/layers/_vectors_layer/model.py:
```diff
@@ -0,0 +1,457 @@
+
```
remove newline at top

kne42 commented on napari/layers/_vectors_layer/model.py:
```diff
@@ -0,0 +1,457 @@
+from typing import Union
+
+import numpy as np
+import scipy.signal as signal
```
```suggestion
from scipy import signal
```

kne42 commented on napari/layers/_vectors_layer/model.py:
```diff
@@ -0,0 +1,457 @@
+from vispy.color import get_color_names
+
+from .view import QtVectorsLayer
+
```
needs two newlines from import statements

kne42 commented on napari/layers/_vectors_layer/model.py:
```diff
@@ -0,0 +1,457 @@
+    Properties
+    ----------
+    vectors : np.ndarray
+        array of shape (N,4) or (N, M , 2)
```
should look more like `vectors : array of shape ...` and include information about what each shape pertains to (instead of what you put in the summary)

kne42 commented on napari/layers/_vectors_layer/model.py:
```diff
@@ -0,0 +1,457 @@
+        array of shape (N,4) or (N, M , 2)
+
+    data : np.ndarray
+        ABC implementation.
```
needs more explicit doc

kne42 commented on napari/layers/_vectors_layer/model.py:
```diff
@@ -0,0 +1,457 @@
+        ABC implementation.
+        returns vectors property setter/getter
+
+    _original_data : np.ndarray
```
do not include private properties anywhere but in the extended summary/notes sections

kne42 commented on napari/layers/_vectors_layer/model.py:
```diff
@@ -0,0 +1,457 @@
+
+    averaging : int
+        (int, int) kernel over which to convolve the data
+        averaging.setter adjusts the underlying data
```
stop referring to setters and getters. properties look to the end user as normal attributes so they may not understand what this means

kne42 commented on napari/layers/_vectors_layer/model.py:
```diff
@@ -0,0 +1,457 @@
+
+    Attributes
+    ----------
+    Private attributes
```
do not include private attributes unless necessary

kne42 commented on napari/layers/_vectors_layer/model.py:
```diff
@@ -0,0 +1,457 @@
+            2) (N, M, 2) array with elements (u, v)
+                where u-v are x-y projections of the vector
+                vector position is one per-pixel in the NxM array
+        :param vectors: ndarray
```
use numpy style docs please

kne42 commented on napari/layers/_vectors_layer/model.py:
```diff
@@ -0,0 +1,457 @@
+        Must preserve data used at construction. Specifically for default
+            averaging/length adjustments
+        averaging/length adjustments recalculate the underlying data
+        :param dat: updated only at construction
```
numpy style docs

kne42 commented on napari/layers/_vectors_layer/model.py:
```diff
@@ -0,0 +1,457 @@
+        :return:
+        """
+
+        if vectors.shape[-1] == 4 and len(vectors.shape) == 2:
```
`len(vectors.shape)` is equivalent to `vectors.ndim`

kne42 commented on napari/layers/_vectors_layer/model.py:
```diff
@@ -0,0 +1,457 @@
+    # ========================= Napari Layer ABC CONTROL methods =====================
+
+
+class InvalidDataFormatError(Exception):
```
special exceptions for the sake of having special exceptions are not necessary

when an argument does not conform to its signature you should raise a `TypeError`

kne42 commented on napari/layers/_vectors_layer/model.py:
```diff
@@ -0,0 +1,457 @@
+
+        # Display vectors if there are any in this slice
+        if len(in_slice_vectors) > 0:
+            data = np.array(in_slice_vectors) + 0.5
```
please explain in a comment as to why you need to add `0.5` to the array

sofroniewn replied:
this is something that happens in the markers layer too to make the beginning of the line appear at the center of the pixel. we should probably explain there too

kne42 commented on napari/layers/_vectors_layer/model.py:
```diff
@@ -0,0 +1,457 @@
+        super().__init__(visual)
+
+        # events for non-napari calculations
+        self.events.add(emit_len=Event,
```
not a big fan of the `emit_` names. would just use what they correspond to, i.e. `length` and `average`

bryantChhun replied:
i brought this up in #182  -- that self.events is actually an EmitterGroup.

But seeing this: https://github.com/vispy/vispy/blob/master/vispy/util/event.py

It looks like vispy likes to define as such:
Canvas.EmitterGroup.EventEmitter

Canvas.events.mouse_press
Canvas.events.resized
Canvas.events.key_press

which is weird to me because the Event itself is one level lower than the EventEmitter... but ok, I can switch back.

kne42 reviewed (commented):
Nice work! Left some comments on documentation and styling.

kne42 committed:
Update napari/layers/_vectors_layer/model.py

Co-Authored-By: bryantChhun <bchhun@gmail.com>

bryantChhun committed:
numpy style docstrings
removed custom exception
some formatting

bryantChhun committed:
numpy style docstrings
removed custom exception

bryantChhun committed:
Merge remote-tracking branch 'origin/Vectors' into Vectors

bryantChhun committed:
reset execution count

bryantChhun committed:
renamed emitters to be "length" or "average"

bryantChhun committed:
also renamed call to events to reflect EventEmitter change

bryantChhun committed:
changes to comments/docs

jni committed:
Merge branch 'master' into Vectors
